### PR TITLE
Add nodeAttributes to RenderMixin

### LINKED ISSUE #22: HTML attributes
I believe @Tomdye has raised this before, but when using `createCachedRenderMixin` their seems to be no other way than completely overriding `getNodeAttributes` for providing HTML attributes other than `classes` and `styles` (which have nearly identical `getters` and `setters`). Perhaps we should have a more generic `attribute` map?

### PR DESCRIPTION
**Pull request by kitsonk**
This PR resolves #22, where expressing attributes in a mixin was challenging.  This uses the new feature of `dojo/compose` that will merge arrays that exist on a prototype (see: dojo/compose#51).  With this PR, a mixin which has the following interface and mixed in with `CachedRenderMixin`:

```typescript
interface NodeAttributeFunction {
    (attributes: VNodeProperties): VNodeProperties;
}

interface NodeAttributeMixin {
    nodeAttributes: NodeAttributeFunction[];
}
```

Will have its node attributes "collected".  The values must be returned from the function in order to be mixed in to the final VNode that is being generated during a render call.

Todo:

- [x] Update unit tests

### CODE REVIEW

**Comment by kitsonk** [comment]
Here is an example of what it would look like, essentially adding it to the compose prototype:

```typescript
const createNodeAttributeMixin = compose({
	nodeAttributes: [ function (this: NodeAttributeMixin): VNodeProperties {
		const props: VNodeProperties = {};

		if (this.type) {
			props['type'] = this.type;
		}
		/* value should always be copied */
		props.value = this.value;
		if ('name' in this.state) {
			props.name = this.state.name;
		}
		if (this.state.disabled) {
			props['disabled'] = 'disabled';
		}

		return props;
	} ]
});
```

**Comment by kitsonk** [comment]
This now requires dojo/compose#59 to be merged and the package published before the CI will be successful.

**Comment by kitsonk** [comment]
Also to note, teasing out several changes to this was getting challenging and so while this PR is larger than ideal, and will cause some downstream refactors, I still feel it is better to just *bite the bullet* on this.

It was becoming evident that the abstraction between `Renderable` and `CachedRenderMixin` was fairly useless and that `Renderable` was more of an abstract interface versus functional code, it made sense to eliminate it and migrate `CachedRenderMixin` to `RenderMixin`.

Also, because of the changes in `compose` around the ordering of init functions, if I had to touch the class because of that, I refactored it so minimise its dependencies and work towards more of a true mixin versus what was somewhat a classical inheritance.  Really now, the most common set of base functionality needed by most mixins is `Stateful` which also provides `Evented` and `Destroyable` and because of the changes in `compose` having to remix these in repeatedly to reorder the init functions is now obsolete.

In another PR though, once we land this one, I would like to refactor the final widgets to fully adopt a more mixin approach and totally eliminate the arbitrary dependency on `createWidget`.

File: src/mixins/createFormFieldMixin.ts [hunk 1]
```diff
@@ -76,52 +102,38 @@ const createFormMixin: FormMixinFactory = compose({
...
 			}
-		}
-	}, (instance: FormField<any>, options: FormFieldMixinOptions<any, FormFieldMixinState<any>>) => {
-		if (options) {
-			const { type } = options;
-			if (type) {
-				instance.type = type;
+		},
+
+		nodeAttributes: [ function (this: FormFieldMixin<any, FormFieldMixinState<any>>): VNodeProperties {
```
Comment (agubler) [inline src/mixins/createFormFieldMixin.ts]: what is the standard for this, we have some examples where `function` is new lined and some inline. I think it reads slightly better new lined by its very subjective.

File: src/mixins/createParentListMixin.ts [hunk 2]
```diff
@@ -82,7 +77,7 @@ const createParentMixin: ParentListMixinFactory = compose<ParentList<Child>, Par
 			}
 		},
 
-		append(this: ParentListMixin<Child>, child: Child | Child[]): Handle {
+		append(this: ParentListMixin<Child>, child: (Child) | (Child)[]): Handle {
```
Comment (agubler) [inline src/mixins/createParentListMixin.ts]: Why the new parentheses?

Comment (kitsonk) [inline src/mixins/createParentListMixin.ts]: Because, strangely, this wasn't working right... It equated to `(Child|Child)[]`, but let me double check that again... we could do `Child[] | Child`... either way it felt ambiguous and I thought it caused some problems.  The problem only highlighted when I passed something that actually didn't match `Child` and was getting argument was unassignable because it didn't contain `.length`.

File: src/mixins/createFormFieldMixin.ts (hunk 1 above)
Comment (kitsonk) [inline src/mixins/createFormFieldMixin.ts]: Hmmm... good point.  I suspect we always haven't been clear.

To try to state it, if the items in the literal can be easily expressed without line breaks, the literal can be expressed without a starting and terminating line break, though with potentially line breaks between the elements.  If all of the elements cannot be expressed without line breaks within the elements, then each element should start and finish with line breaks, including the leading and terminating line break for the literal.

TL;DR I agree with you and will change it.

File: tests/unit/mixins/createRenderMixin.ts [hunk 3]
```diff
@@ -97,14 +90,14 @@ registerSuite({
...
+			assert.include(id, createRenderMixin.idBase, 'should include static idBase');
+			assert.notStrictEqual(cachedRender.id, createRenderMixin.idBase, 'but shouldn\'t match exactly');
 			assert.strictEqual(id, cachedRender.id, 'should return same id');
 		},
 		'is read only'() {
-			const cachedRender = createCachedRenderMixin();
+			const cachedRender = createRenderMixin();
 			assert.throws(() => {
-				cachedRender.id = 'foo';
+				(<any> cachedRender).id = 'foo';
```
Comment (kitsonk) [inline tests/unit/mixins/createRenderMixin.ts]: Because `.id` is marked as `readonly` now, which it is at runtime, but without `any`, TypeScript tries to keep you from making that mistake.

File: tests/unit/mixins/createRenderMixin.ts [hunk 4]
```diff
@@ -46,22 +38,23 @@ registerSuite({
...
 		assert.deepEqual(cachedRender.getChildrenNodes(), [ 'foo' ]);
 	},
 	'render()/invalidate()'() {
-		const cachedRender = createCachedRenderMixin({
+		const cachedRender = createRenderMixin({
 			state: { id: 'foo', label: 'foo' }
 		});
-		cachedRender.invalidate();
-		cachedRender.invalidate();
+		// cachedRender.on('statechange', () => cachedRender.invalidate());
```
Comment (kitsonk) [inline tests/unit/mixins/createRenderMixin.ts]: Fixed.

File: tests/functional/widgetClasses.ts [hunk 5]
```diff
@@ -45,6 +45,8 @@ addClassButton.on('click', (e: MouseEvent) => {
 	header.setState(nextState);
 });
 
+console.log(addClassButton);
```
Comment (kitsonk) [inline tests/functional/widgetClasses.ts]: Fixed.

File: tests/functional/tabbedPanel.html [hunk 6]
```diff
@@ -25,10 +25,10 @@
 			],
 			map: {
```
Comment (kitsonk) [inline tests/functional/tabbedPanel.html]: fixed

File: tests/unit/mixins/createRenderMixin.ts (hunk 3 above)
Comment (agubler) [inline tests/unit/mixins/createRenderMixin.ts]: ah yeah that makes sense.

File: src/mixins/createRenderMixin.ts [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+	 * @param overrides Any optional overrides of properties
+	 */
+	getNodeAttributes(overrides?: VNodeProperties): VNodeProperties;
+
+	/**
+	 * Returns any children VNodes that are part of the widget
+	 */
+	getChildrenNodes(): (VNode | string)[];
+
+	readonly id: string;
```
Comment (kitsonk) [inline src/mixins/createRenderMixin.ts]: fixed

File: tests/unit/mixins/createRenderMixin.ts (hunk 3 above)
Comment (kitsonk) [inline tests/unit/mixins/createRenderMixin.ts]: I did add a comment to explain.

File: src/mixins/createParentListMixin.ts (hunk 2 above)
Comment (kitsonk) [inline src/mixins/createParentListMixin.ts]: I don't know what I was thinking, because it doesn't break anything when I take it out... must have been a moment of madness...  Anyways, when for `Child[] | Child`.

File: src/mixins/createRenderMixin.ts [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+				return (this.state && this.state.id) || generateID(this);
+			},
+
+			invalidate(this: RenderMixin<RenderMixinState>): void {
+				if (dirtyMap.get(this)) { /* short circuit if already dirty */
+					return;
+				}
+				const parent = this.parent;
+				dirtyMap.set(this, true);
+				renderCache.delete(this); /* Allow GC to occur on renderCache */
```
Comment (agubler) [inline src/mixins/createRenderMixin.ts]: do we need to delete on a `weakMap`?

File: src/mixins/createVNodeEvented.ts [hunk 9]
```diff
@@ -1,37 +1,48 @@
...
-	onmouseenter?(ev?: MouseEvent): boolean | void;
-	onmouseleave?(ev?: MouseEvent): boolean | void;
-	onmousemove?(ev?: MouseEvent): boolean | void;
-	onmouseout?(ev?: MouseEvent): boolean | void;
-	onmouseover?(ev?: MouseEvent): boolean | void;
-	onmouseup?(ev?: MouseEvent): boolean | void;
-	onmousewheel?(ev?: MouseWheelEvent): boolean | void;
-	onscroll?(ev?: UIEvent): boolean | void;
-	onsubmit?(ev?: Event): boolean | void;
+	[on: string]: (ev?: TargettedEventObject) => boolean | undefined | null;
```
Comment (agubler) [inline src/mixins/createVNodeEvented.ts]: probably cleaner if we have a type for the event returns

File: src/mixins/createRenderMixin.ts (hunk 8 above)
Comment (kitsonk) [inline src/mixins/createRenderMixin.ts]: In this situation, the instance will persist and therefore the cache will persist until the cache is set again (which may never occur)... Therefore an invalidated cache would not be freed up for GC until a) the instance is dereferenced or b) the widget is rendered again.  Assuming you had a large part of an application that is not visible but has been invalidated, that could be quite a large cache which you could free up for GC.

File: src/mixins/createVNodeEvented.ts (hunk 9 above)
Comment (kitsonk) [inline src/mixins/createVNodeEvented.ts]: Good idea, will address.

File: src/mixins/createRenderMixin.ts (hunk 8 above)
Comment (agubler) [inline src/mixins/createRenderMixin.ts]: makes sense
